kmall.py

```diff
@@ -3425,37 +3481,201 @@ def seek_next_startbyte(self, file_length, start_ptr=0):
                     self.FID.seek(possible_start - 4)
                     return True
 
-    def sequential_read_records(self, first_installation_rec=False, start_ptr=0, end_ptr=0):
-        ########### STILL IN DEVELOPMENT ############
-        raise NotImplementedError('sequential_read_records is still in development')
-
-        # data we are reading
-        # systemID is not the same thing as serial number, only seems to be in the IIP
-        recs_categories = {'SKM': ['header.dgtime', 'sample.KMdefault.roll_deg', 'sample.KMdefault.pitch_deg',
-                                   'sample.KMdefault.heave_m', 'sample.KMdefault.heading_deg',
-                                   'sample.KMdefault.latitude_deg', 'sample.KMdefault.longitude_deg',
-                                   'sample.KMdefault.ellipsoidHeight_m'],
-                           'IIP': ['header.dgtime', 'install_txt'],
-                           'MRZ': ['header.dgtime', 'cmnPart.pingCnt', 'pingInfo.soundSpeedAtTxDepth_mPerSec',
-                                   'pingInfo.numTxSectors', 'header.systemID', 'txSectorInfo.txSectorNumb',
-                                   'txSectorInfo.tiltAngleReTx_deg', 'txSectorInfo.sectorTransmitDelay_sec',
-                                   'txSectorInfo.centreFreq_Hz', 'sounding.beamAngleReRx_deg',
-                                   'sounding.txSectorNumb', 'sounding.detectionType', 'sounding.qualityFactor', 
-                                   'sounding.twoWayTravelTime_sec', 'pingInfo.modeAndStabilisation',
-                                   'pingInfo.pulseForm', 'pingInfo.depthMode'],
-                           'SVP': ['time_sec', 'sensorData.depth_m', 'sensorData.soundVelocity_mPerSec']}
-        
+    def _divide_rec(self, rec):
+        """
+        MRZ comes in from sequential read by time/ping.  Each ping may have multiple sectors to it which we want
+        to treat as separate pings.  Do this by generating a new record for each sector in the ping.
+        """
+        if self.datagram_ident != 'MRZ':
+            return [rec]
+        elif rec['pingInfo']['numTxSectors'] == 1:
+            return [rec]
+        else:
+            totalrecs = []
+            pingtime = rec['header']['dgtime']
+            for sec in rec['txSectorInfo']['txSectorNumb']:
+                split_rec = copy.copy(rec)
```

Comment:
Do you think it is better to use `copy` or `deepcopy` here? Certainly less memory intensive to use copy. But copy is kind of counter intuitive since changes to the copy change the original. If you keep `copy`, maybe make a note in the doc string that the new records contain references to the original record so there's some hope that I won't do something stupid and get something unexpected.

Reply:
A shallow copy is what we want I think.  I had an issue with changing the time stamp that led me to deepcopy at first, but the whole process takes 10x longer with deepcopy (my test went from 2 secs to about 20 secs)